src/fabric_quality_factor.erl

```diff
@@ -0,0 +1,213 @@
+    DbName = mem3:dbname(ShardName),
+    Nodes = nodes_for_shard(ShardName, DbName),
+
+    %% For each TargetNode with a ShardName copy, find the upper bound of
+    %% updates to other shard replicas which are missing from TargetNode
+    MissingUpdates = [shard_sync_quality_factor(ShardName, TargetNode) ||
+        TargetNode <- Nodes],
+
+    %% The quality factor is the sum of possible missing updates
+    {ok, lists:sum(MissingUpdates)}.
+
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+%% Internal functions for QF determination
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+
+%% Calculate the shard sync quality factor for a single node/shard copy. This
+%% is the number of updates to other shard copies not present on TargetNode
+%% ShardName  - e.g. <<"shards/c0000000-ffffffff/ksnavely/motohacker.1368999314">>
+%% TargetNode - e.g. 'dbcore@db11.julep.cloudant.net'
+shard_sync_quality_factor(not_found, _TargetNode) ->
+    not_found;
+shard_sync_quality_factor(ShardName, TargetNode) ->
+    DbName = mem3:dbname(ShardName),
+    Nodes = nodes_for_shard(ShardName, DbName),
+
+    %% Find the upper bound of updates to other shard copies which are missing
+    %% from TargetNode
+    MissingUpdates = [shard_missing_updates(Source, TargetNode, ShardName) ||
+        Source <- Nodes, Source =/= TargetNode],
+
+    %% The quality factor is the sum of possible missing updates
+    lists:sum(MissingUpdates).
+
+%% Fetch a single checkpoint doc given target, source, and shard name
+%% ShardName  - e.g. <<"shards/c0000000-ffffffff/ksnavely/motohacker.1368999314">>
+%% TargetNode - e.g. 'dbcore@db14.julep.cloudant.net'
+%% SourceNode - e.g. 'dbcore@db14.julep.cloudant.net'
+local_checkpoint(TargetNode, SourceNode, ShardName) ->
+    {ok, Db} = rpc:call(SourceNode, couch_db, open, [ShardName, [?ADMIN_CTX]]),
+    UUID = couch_db:get_uuid(Db),
+    {_LocalId, Doc} = mem3_rpc:load_checkpoint(TargetNode, ShardName, SourceNode, UUID),
+    Doc.
+
+%% Find upper bound of updates to ShardName on SourceNode that are missing
+%% from TargetNode, which also holds a copy of ShardName
+%% SourceNode - e.g. 'dbcore@db9.julep.cloudant.net'
+%% TargetNode - e.g. 'dbcore@db11.julep.cloudant.net'
+%% ShardName  - e.g. <<"shards/c0000000-ffffffff/ksnavely/motohacker.1368999314">>
+shard_missing_updates(SourceNode, TargetNode, ShardName) ->
+    %% The number of missing updates is the difference in sequence numbers.
+    shard_seq(SourceNode, ShardName) - shard_checkpoint_seq(SourceNode, TargetNode, ShardName).
+
+%% Fetch the last replicated source seq from the replication checkpoint doc.
+%% SourceNode - e.g. 'dbcore@db9.julep.cloudant.net'
+%% TargetNode - e.g. 'dbcore@db14.julep.cloudant.net'
+%% ShardName  - e.g. <<"shards/c0000000-ffffffff/ksnavely/motohacker.1368999314">>
+shard_checkpoint_seq(SourceNode, TargetNode, ShardName) ->
+    CheckpointDoc = local_checkpoint(TargetNode, SourceNode, ShardName),
+    {Body} = element(4, CheckpointDoc),
```

Comment:
Referring to element positions directly can make for some brittle code. I'd suggest using the record syntax `CheckpointDoc#doc.body` instead.

Now, that can also be brittle in a clustered system with online upgrades, so we've been moving to implement accessors for records that are shared across modules. Over time we'd want to see something like `{Body} = couch_doc:get_body(CheckpointDoc)` here.